### DISCUSSION TITLE
How to use SignerCap?

### DISCUSSION THREAD

**Discussion by memxor** [aptos-labs/aptos-developer-discussions#54] [Questions] [answered]
I'm having a hard time understanding how to create a SignerCap, this is the code
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/9eb5ccae-ef01-4db5-bcd0-7c636b30bc12)

I did create a resource account from the terminal, using the same address as the publisher, but I'm getting the following error - `"Error": "Simulation failed with status: Move abort in 0x1::resource_account: 0x10002"`

**Comment by lopeselio** [comment]
@gregnazario Hi Greg, would appreciate any help on this. 
It says requires a resource account, we did create one but its not working! We're working on the smart contract integrations for our game

**Comment by MoonShiesty** [comment]
@memxor can you confirm that the @publisher account is the account which publishes the resource_account?

**Comment by MoonShiesty** [comment]
@memxor i believe i see the issue. the publisher account needs to be a different account than the resource_account. can you try creating a different account, such as @deployer to publish the @publisher resource account?

for example: https://github.com/aptos-labs/aptos-core/blob/2e9d8ee759fcd3f6e831034f05c1656b1c48efc4/aptos-move/move-examples/mint_nft/sources/minting.move#L83

**Reply by lopeselio** [reply]
Thanks @MoonShiesty we will check it out

**Reply by MoonShiesty** [reply]
youll have to change the @publisher address in the Move.toml file. you can generate the new resource_account address using:

```
aptos account derive-resource-account-address
```

**Reply by memxor** [reply]
Am I suppose to use this `Result` address somewhere? 
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/2b256397-cf9e-42c4-a466-1220a72a1c48)

**Reply by memxor** [reply]
You mean I turn the current `publisher` into a `resource_publisher` and have a new address for `publisher`?

**Reply by MoonShiesty** [reply]
yep, youll want to set publisher in the Move.toml file to the generated resource-account address

**Reply by MoonShiesty** [reply]
you need 2 different accounts. a @publisher (thats funded with gas) and the resource account object @minter (any different name works), where the package is published to

**Reply by memxor** [reply]
I created a new account and added that as publisher in Move.toml, but now I'm getting this

![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/28e93080-a955-4c76-97b9-c407a54a4887)

It says the address dont match

**Reply by MoonShiesty** [reply]
i think this is because you set @publisher to two different addresses, either in your `Move.toml` file, or through `--named-addresses`

**Reply by memxor** [reply]
No, sorry I don't understand. I've `0xd330fae5fcb5b13ebe7cf9a8058bfaef0381bf76a9167b1c84635ee724dacb6c` which published the resource_account and is marked as `resource_signer` in Move.toml. And I've `0x84f2754bd9966607271c8124fc43738ea0642822b40ebda05970c4ab52d16991` marked as publisher, which is also in `config.yaml` 

Then I'm running `aptos move publish`

![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/452fca96-46a2-4520-b492-41f37b2cf21d)
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/5ddd3212-5f77-424d-a2ed-41b3479e21a7)

**Comment by MoonShiesty** [comment, marked as answer]
can you try:

``` aptos move create-resource-account-and-publish-package --seed MY_SEED --address-name publisher```

also believe this cli command pass in the named address `publisher`, so you can wildcard `publisher = "_"` in the `[addresses]` section (you'll still need it in `[dev-addresses])`, but the way you did it should also work

**Reply by memxor** [reply]
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/7b8d2616-7c38-42a6-8785-fe6de88b16c2)
It says it already exists

**Reply by MoonShiesty** [reply]
ah, you must have already created the resource account. try a new seed or resource_publisher account/address

**Reply by memxor** [reply]
Why I need to create another one, if I already have one?

**Reply by MoonShiesty** [reply]
i believe the aptos-framework supports this, just sure  where/if that feature exists in the cli. have to defer to @gregnazario on how to publish a package to an already existing resource account

if this is blocking you, give `create-resource-account-and-publish-package` a try with a new publisher account/seed

**Reply by memxor** [reply]
A/C to this error 
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/bb67b042-cae0-4472-9a7d-21248841782c)
from this if I take `0x3d08df89d946c628f742964011f87915f1b22502f385888f585d97a8c7e3c305` as my `resource_publisher` in Move.toml, it should work right?

But then I get this 
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/d886ca59-ff53-490c-b6cc-00b67e8f6955)

**Reply by MoonShiesty** [reply]
can you paste the `Move.toml` file?

you need to update `resource_publisher` address if you updated the account deploying the publisher module
and `publisher` addresses so it reflects the resource account address generated from new publisher address/seed

**Reply by memxor** [reply]
```
[package]
name = "Horse_Race_Aptos"
version = "1.0.0"
license = "MIT"
authors = []

[addresses]
resource_publisher = "0x101510bb574fbdd156d267b4832fe8bf461836f97681b797adac7720fc26f7e4"
publisher = "0x47247b7075b514ae5c8ab3b442ff601ebad12b9b9e3fc41c3f4ef4ed88c0094e"

[dev-addresses]
publisher = "0x47247b7075b514ae5c8ab3b442ff601ebad12b9b9e3fc41c3f4ef4ed88c0094e"

[dependencies.AptosFramework]
git = "https://github.com/aptos-labs/aptos-core.git"
rev = "randomnet"
subdir = "aptos-move/framework/aptos-framework"

[dependencies.AptosTokenObjects]
git = "https://github.com/aptos-labs/aptos-core.git"
rev = "randomnet"
subdir = "aptos-move/framework/aptos-token-objects"
```

**Reply by memxor** [reply]
maybe it is too much to ask, but can try deploying it? https://github.com/IntoTheVerse/APTHorse-contracts

**Reply by MoonShiesty** [reply]
just deployed it to devnet, steps should be the same for randomnet:

```
aptos move create-resource-account-and-publish-package --seed MY_SEED --address-name publisher
```
https://explorer.aptoslabs.com/account/0xf08743ba90ed781dc21ecf8998cbbe86e8af8ad1687f638ecc3b9db21390b85f/modules/code/aptos_horses?network=devnet

i changed the move.toml file as follows:

```
[addresses]
resource_publisher = "0x49c519a35b0c94a1825a1e493fdf6f2a3c5c3ad03ca1803327947d2f7740af78"
publisher = "_"

[dev-addresses]
publisher = "0xcafe"
```

just update `resource_publisher` and `MY_SEED` to your resource publisher account/seed, and you should be good to go

the `[dev_address]` change isn't required, but you can compile with the `--dev` flag now and not worry about setting that value

**Reply by memxor** [reply]
I did `aptos move create-resource-account-and-publish-package --seed MY_SEED --address-name publisher` which gave me this
![image](https://github.com/aptos-labs/aptos-developer-discussions/assets/48633453/b3282e1d-a29d-4636-9813-3bcbd291c938)

Should I use `0x801ffe293d57933a5f6f9886aa051ba52d39d4190fa7113e9d7d6fcc5681b2b3` as the `publisher` or `resource_publisher`?

**Reply by MoonShiesty** [reply]
nice, looks like it works!

`0x801ffe293d57933a5f6f9886aa051ba52d39d4190fa7113e9d7d6fcc5681b2b3`  is the address of the new resource account the new `publisher` module is published under. if another module depends on `publisher` they'll need to use that address in their `Move.toml`/`named-addresses`.

 `resource_publisher` is the account who created the resource account and module and paid the gas.

**Comment by JohnChangUK** [comment]
Thanks @MoonShiesty for helping out. It looks like this person is trying to use resource accounts to deploy contracts? Here is a tutorial I wrote for this: https://github.com/aptos-labs/developer-docs/blob/main/apps/docusaurus/docs/tutorials/programmatic-upgradeable-module.md

Soon in mainnet, there will be a way to deploy code to objects, which makes the entire user experience much simpler. Will update the community on when that'll happen.

**Reply by MoonShiesty** [reply]
> Soon in mainnet, there will be a way to deploy code to objects, which makes the entire user experience much simpler. 

im really excited for this because we use resource accounts to deploy code. 

but we don't only use the `resource_account` `signer_cap` for package upgrades. we also use it to generate signers for managing owned objects (Tokens, FungibleStores). similar to how the user in the linked issue uses the `resource_account` `signer_cap` to transfer ownership of an NFT from the creator (resource account module) to user. the aptos-framework provides a relative easy-to-use pattern to do this with `resource_account::retrieve_resource_account_cap`.

looking at the module object framework, its not clear to me how objects can use the resource account to generate signatures for things like managing objects. if there's no straight-forward way to sign with object modules we'd have to continue using resource accounts.
